cpu/sources/layout7/cpu_constraint_poly.move

```diff
@@ -0,0 +1,3926 @@
+        let initial_poseidon_addr = *borrow(&ctx, 32);
+        let trace_generator = *borrow(&ctx, 33);
+        let oods_point = *borrow(&ctx, 34);
+
+        let point = oods_point;
+
+        {
+            // compute expmods
+            // expmods[0] = point^(trace_length / 2048)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 2048));
+                *borrow_mut(&mut ctx, 286) = val;
+            };
+            // expmods[1] = point^(trace_length / 1024)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 1024));
+                *borrow_mut(&mut ctx, 287) = val;
+            };
+            // expmods[2] = point^(trace_length / 128)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 128));
+                *borrow_mut(&mut ctx, 288) = val;
+            };
+            // expmods[3] = point^(trace_length / 64)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 64));
+                *borrow_mut(&mut ctx, 289) = val;
+            };
+            // expmods[4] = point^(trace_length / 32)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 32));
+                *borrow_mut(&mut ctx, 290) = val;
+            };
+            // expmods[5] = point^(trace_length / 16)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 16));
+                *borrow_mut(&mut ctx, 291) = val;
+            };
+            // expmods[6] = point^(trace_length / 4)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 4));
+                *borrow_mut(&mut ctx, 292) = val;
+            };
+            // expmods[7] = point^(trace_length / 2)
+            {
+                let val = fpow(/*point*/ point, (/*trace_length*/ trace_length / 2));
+                *borrow_mut(&mut ctx, 293) = val;
+            };
+            // expmods[8] = point^trace_length
+            {
+                let val = fpow(/*point*/ point, /*trace_length*/ trace_length);
+                *borrow_mut(&mut ctx, 294) = val;
+            };
+            // expmods[9] = trace_generator^(trace_length / 64)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (/*trace_length*/ trace_length / 64));
+                *borrow_mut(&mut ctx, 295) = val;
+            };
+            // expmods[10] = trace_generator^(trace_length / 32)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (/*trace_length*/ trace_length / 32));
+                *borrow_mut(&mut ctx, 296) = val;
+            };
+            // expmods[11] = trace_generator^(3 * trace_length / 64)
+            {
+                let val = fpow(/*trace_generator*/ trace_generator, (fmul(3, /*trace_length*/ trace_length) / 64));
```

Comment:
Pre-compute `tg64 = trace_generator^(trace_length / 64)`, then `expmods[11] = fpow(tg64, 3)`. So we can reuse `tg64`.